drivers/virtio/virtio_mmio.c

```diff
@@ -0,0 +1,419 @@
+	struct k_spinlock notify_lock;
+};
+
+struct virtio_mmio_config {
+	DEVICE_MMIO_NAMED_ROM(reg_base);
+};
+
+static inline uint32_t virtio_mmio_read32(const struct device *dev, uint32_t offset)
+{
+	const mem_addr_t reg = DEVICE_MMIO_NAMED_GET(dev, reg_base) + offset;
+	uint32_t val;
+
+	barrier_dmem_fence_full();
+	val = sys_read32(reg);
+	return sys_le32_to_cpu(val);
+}
+
+static inline void virtio_mmio_write32(const struct device *dev, uint32_t offset, uint32_t val)
+{
+	const mem_addr_t reg = DEVICE_MMIO_NAMED_GET(dev, reg_base) + offset;
+
+	sys_write32(sys_cpu_to_le32(val), reg);
+	barrier_dmem_fence_full();
+}
+
+static void virtio_mmio_isr(const struct device *dev)
+{
+	struct virtio_mmio_data *data = dev->data;
+	struct virtq *virtqs = data->virtqueues;
+	k_spinlock_key_t key = k_spin_lock(&data->isr_lock);
+
+	const uint32_t isr_status = virtio_mmio_read32(dev, VIRTIO_MMIO_INTERRUPT_STATUS);
+
+	virtio_mmio_write32(dev, VIRTIO_MMIO_INTERRUPT_ACK, isr_status);
+
+	if (isr_status & VIRTIO_MMIO_QUEUE_INTERRUPT) {
+		for (int i = 0; i < data->virtqueue_count; i++) {
+			const uint16_t used_idx = sys_le16_to_cpu(virtqs[i].used->idx);
+
+			while (virtqs[i].last_used_idx != used_idx) {
+				const uint16_t idx = virtqs[i].last_used_idx % virtqs[i].num;
+				const uint16_t idx_le = sys_cpu_to_le16(idx);
+				const uint16_t chain_head_le = virtqs[i].used->ring[idx_le].id;
+				const uint16_t chain_head = sys_le16_to_cpu(chain_head_le);
+				const uint32_t used_len =
+					sys_le32_to_cpu(virtqs[i].used->ring[idx_le].len);
+
+				/*
+				 * We are making a copy here, because chain will be
+				 * returned before invoking the callback and may be
+				 * overwritten by the time callback is called. This
+				 * is to allow callback to immediately place the
+				 * descriptors back in the avail_ring
+				 */
+				struct virtq_receive_callback_entry cbe =
+					virtqs[i].recv_cbs[chain_head];
+
+				uint16_t next = chain_head;
+				bool last = false;
+
+				/*
+				 * We are done processing the descriptor chain, and
+				 * we can add used descriptors back to the free stack.
+				 * The only thing left to do is calling the callback
+				 * associated with the chain, but it was saved above on
+				 * the stack, so other code is free to use the descriptors
+				 */
+				while (!last) {
+					const uint16_t curr = next;
+					const uint16_t curr_le = sys_cpu_to_le16(curr);
+
+					next = virtqs[i].desc[curr_le].next;
+					last = !(virtqs[i].desc[curr_le].flags & VIRTQ_DESC_F_NEXT);
+					virtq_add_free_desc(&virtqs[i], curr);
+				}
+
+				virtqs[i].last_used_idx++;
+
+				if (cbe.cb) {
+					cbe.cb(cbe.opaque, used_len);
+				}
+			}
+		}
+	}
+	if (isr_status & VIRTIO_MMIO_DEVICE_CONFIGURATION_INTERRUPT) {
+		LOG_ERR("device configuration change interrupt is currently unsupported");
+	}
+
+	k_spin_unlock(&data->isr_lock, key);
+}
+
+static void virtio_mmio_reset(const struct device *dev)
+{
+	virtio_mmio_write32(dev, VIRTIO_MMIO_STATUS, 0);
```

Comment:
Probably none of the VMMs enter VM again before setting VIRTIO status to 0 (i.e. mid-reset), but the spec (2.4.2) states that reset should be considered complete when we read 0 from status, so it might be better to wait until reading status returns 0 before returning from this function.